<a href="https://colab.research.google.com/github/balu337/LendingClubRepo/blob/master/LC_v1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **ImportStatments**

In [112]:
# Numerical and Data Analysis
import numpy as np
import pandas as pd

# Data Visualization
import seaborn as sns
import matplotlib.pyplot as plt

# Extra
import warnings
warnings.filterwarnings('ignore')
pd.set_option('display.max_rows',200)

In [113]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [114]:
filepath="/content/drive/MyDrive/IITB_AIML/1_Statistics/LendingClubCaseStudy/loan.csv"
#filepath="/content/drive/MyDrive/LendingClub/loan.csv"


df = pd.read_csv(filepath)

## **Analyze Raw Data**

Check column by column using Data Dictionary and get unerstanding the data ( content and type )

In [115]:
#Dataset Shape
df.shape

(39717, 111)

In [116]:
#Description of dataset
df.describe()

id     member_id     loan_amnt   funded_amnt  \
count  3.971700e+04  3.971700e+04  39717.000000  39717.000000   
mean   6.831319e+05  8.504636e+05  11219.443815  10947.713196   
std    2.106941e+05  2.656783e+05   7456.670694   7187.238670   
min    5.473400e+04  7.069900e+04    500.000000    500.000000   
25%    5.162210e+05  6.667800e+05   5500.000000   5400.000000   
50%    6.656650e+05  8.508120e+05  10000.000000   9600.000000   
75%    8.377550e+05  1.047339e+06  15000.000000  15000.000000   
max    1.077501e+06  1.314167e+06  35000.000000  35000.000000   

       funded_amnt_inv   installment    annual_inc           dti  \
count     39717.000000  39717.000000  3.971700e+04  39717.000000   
mean      10397.448868    324.561922  6.896893e+04     13.315130   
std        7128.450439    208.874874  6.379377e+04      6.678594   
min           0.000000     15.690000  4.000000e+03      0.000000   
25%        5000.000000    167.020000  4.040400e+04      8.170000   
50%        8975.000000    280.220000  5.900000e+04     13.400000   
75%       14400.000000    430.780000  8.230000e+04     18.600000   
max       35000.000000   1305.190000  6.000000e+06     29.990000   

        delinq_2yrs  inq_last_6mths  ...  num_tl_90g_dpd_24m  \
count  39717.000000    39717.000000  ...                 0.0   
mean       0.146512        0.869200  ...                 NaN   
std        0.491812        1.070219  ...                 NaN   
min        0.000000        0.000000  ...                 NaN   
25%        0.000000        0.000000  ...                 NaN   
50%        0.000000        1.000000  ...                 NaN   
75%        0.000000        1.000000  ...                 NaN   
max       11.000000        8.000000  ...                 NaN   

       num_tl_op_past_12m  pct_tl_nvr_dlq  percent_bc_gt_75  \
count                 0.0             0.0               0.0   
mean                  NaN             NaN               NaN   
std                   NaN             NaN               NaN   
min                   NaN             NaN               NaN   
25%                   NaN             NaN               NaN   
50%                   NaN             NaN               NaN   
75%                   NaN             NaN               NaN   
max                   NaN             NaN               NaN   

       pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  total_bal_ex_mort  \
count          39020.000000    39678.0              0.0                0.0   
mean               0.043260        0.0              NaN                NaN   
std                0.204324        0.0              NaN                NaN   
min                0.000000        0.0              NaN                NaN   
25%                0.000000        0.0              NaN                NaN   
50%                0.000000        0.0              NaN                NaN   
75%                0.000000        0.0              NaN                NaN   
max                2.000000        0.0              NaN                NaN   

       total_bc_limit  total_il_high_credit_limit  
count             0.0                         0.0  
mean              NaN                         NaN  
std               NaN                         NaN  
min               NaN                         NaN  
25%               NaN                         NaN  
50%               NaN                         NaN  
75%               NaN                         NaN  
max               NaN                         NaN  

[8 rows x 87 columns]

In [117]:
#Display information about dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Columns: 111 entries, id to total_il_high_credit_limit
dtypes: float64(74), int64(13), object(24)
memory usage: 33.6+ MB


In [118]:
#Find datatypes of columns
df.dtypes

id                                  int64
member_id                           int64
loan_amnt                           int64
funded_amnt                         int64
funded_amnt_inv                   float64
term                               object
int_rate                           object
installment                       float64
grade                              object
sub_grade                          object
emp_title                          object
emp_length                         object
home_ownership                     object
annual_inc                        float64
verification_status                object
issue_d                            object
loan_status                        object
pymnt_plan                         object
url                                object
desc                               object
purpose                            object
title                              object
zip_code                           object
addr_state                        

In [119]:
#Get All Column Names

col= list(df.columns)
print("# Columns are ", len(col))
print("Columns are ", col)

# Columns are  111
Columns are  ['id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv', 'term', 'int_rate', 'installment', 'grade', 'sub_grade', 'emp_title', 'emp_length', 'home_ownership', 'annual_inc', 'verification_status', 'issue_d', 'loan_status', 'pymnt_plan', 'url', 'desc', 'purpose', 'title', 'zip_code', 'addr_state', 'dti', 'delinq_2yrs', 'earliest_cr_line', 'inq_last_6mths', 'mths_since_last_delinq', 'mths_since_last_record', 'open_acc', 'pub_rec', 'revol_bal', 'revol_util', 'total_acc', 'initial_list_status', 'out_prncp', 'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee', 'recoveries', 'collection_recovery_fee', 'last_pymnt_d', 'last_pymnt_amnt', 'next_pymnt_d', 'last_credit_pull_d', 'collections_12_mths_ex_med', 'mths_since_last_major_derog', 'policy_code', 'application_type', 'annual_inc_joint', 'dti_joint', 'verification_status_joint', 'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m', 'open

### Creating new dataframe for Analysis

In [120]:
#Removing unwanted columns

# #After basic analysis columns after [ annual_inc_joint ] are not useful as there is not data.

# print("-->",col[53])
# columns_removed=col[53:]
# print("removed col ",columns_removed)
# df1=df.drop(columns_removed,axis=1)

# Lets drop columns with all null values
df1=df.dropna(axis=1, how='all')
df1.shape



(39717, 57)

In [121]:
# If we observe there are few columns have sigle values, those are not helpful for analysis lets drop them

#check how many single value columns are there
# print((df1.nunique()==1).sum())
# print(df1[nunique()==1])
un=df1.nunique()==1
single_unique_cols = list( un[un].index )
print(single_unique_cols)



['pymnt_plan', 'initial_list_status', 'collections_12_mths_ex_med', 'policy_code', 'application_type', 'acc_now_delinq', 'chargeoff_within_12_mths', 'delinq_amnt', 'tax_liens']


In [122]:
df1.drop(columns= single_unique_cols, inplace=True)

#Print Shape
df1.shape

(39717, 48)

In [123]:
#Get  Column afte dropping columns after annual_inc_joint

col1= list(df1.columns)
print("# Columns are ", len(col1))
print("Columns are ", col1)




# Columns are  48
Columns are  ['id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv', 'term', 'int_rate', 'installment', 'grade', 'sub_grade', 'emp_title', 'emp_length', 'home_ownership', 'annual_inc', 'verification_status', 'issue_d', 'loan_status', 'url', 'desc', 'purpose', 'title', 'zip_code', 'addr_state', 'dti', 'delinq_2yrs', 'earliest_cr_line', 'inq_last_6mths', 'mths_since_last_delinq', 'mths_since_last_record', 'open_acc', 'pub_rec', 'revol_bal', 'revol_util', 'total_acc', 'out_prncp', 'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee', 'recoveries', 'collection_recovery_fee', 'last_pymnt_d', 'last_pymnt_amnt', 'next_pymnt_d', 'last_credit_pull_d', 'pub_rec_bankruptcies']


In [124]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Data columns (total 48 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       39717 non-null  int64  
 1   member_id                39717 non-null  int64  
 2   loan_amnt                39717 non-null  int64  
 3   funded_amnt              39717 non-null  int64  
 4   funded_amnt_inv          39717 non-null  float64
 5   term                     39717 non-null  object 
 6   int_rate                 39717 non-null  object 
 7   installment              39717 non-null  float64
 8   grade                    39717 non-null  object 
 9   sub_grade                39717 non-null  object 
 10  emp_title                37258 non-null  object 
 11  emp_length               38642 non-null  object 
 12  home_ownership           39717 non-null  object 
 13  annual_inc               39717 non-null  float64
 14  verification_status   

In [125]:
df1.isnull().mean()*100

id                          0.000000
member_id                   0.000000
loan_amnt                   0.000000
funded_amnt                 0.000000
funded_amnt_inv             0.000000
term                        0.000000
int_rate                    0.000000
installment                 0.000000
grade                       0.000000
sub_grade                   0.000000
emp_title                   6.191303
emp_length                  2.706650
home_ownership              0.000000
annual_inc                  0.000000
verification_status         0.000000
issue_d                     0.000000
loan_status                 0.000000
url                         0.000000
desc                       32.585543
purpose                     0.000000
title                       0.027696
zip_code                    0.000000
addr_state                  0.000000
dti                         0.000000
delinq_2yrs                 0.000000
earliest_cr_line            0.000000
inq_last_6mths              0.000000
m

Dropping colmns with null values more than 40%.

**Columns                   % of Null Values**


1.   mths_since_last_delinq     64.662487
2.   mths_since_last_record     92.985372
3.   next_pymnt_d               97.129693







In [126]:
null_percentage = df1.isnull().mean() * 100

# Identify columns where the percentage of null values is greater than 40%
columns_to_drop = null_percentage[null_percentage > 40].index

# Drop these columns from the DataFrame
df1 = df1.drop(columns=columns_to_drop)



In [127]:
#Finding Shape after removig columns with null > 40%
df1.shape
df1.isnull().mean()*100

id                          0.000000
member_id                   0.000000
loan_amnt                   0.000000
funded_amnt                 0.000000
funded_amnt_inv             0.000000
term                        0.000000
int_rate                    0.000000
installment                 0.000000
grade                       0.000000
sub_grade                   0.000000
emp_title                   6.191303
emp_length                  2.706650
home_ownership              0.000000
annual_inc                  0.000000
verification_status         0.000000
issue_d                     0.000000
loan_status                 0.000000
url                         0.000000
desc                       32.585543
purpose                     0.000000
title                       0.027696
zip_code                    0.000000
addr_state                  0.000000
dti                         0.000000
delinq_2yrs                 0.000000
earliest_cr_line            0.000000
inq_last_6mths              0.000000
o

### Questions

In [128]:
# Do we need Emp title ?


## **Cleansing Data**


* Find Shape of data
* Check for Null columns and cells
* For numeric data use either median ( mostly ) or mode as required
* For categorical data use Mode

** Remove Unwanted colums**

In [129]:
df1 = df1.drop(columns=['desc','title','url'])

 **Fix the Fromat of the columns by conversion where its required. e,g Emp_lenght need to be converted to int**

In [130]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Data columns (total 42 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       39717 non-null  int64  
 1   member_id                39717 non-null  int64  
 2   loan_amnt                39717 non-null  int64  
 3   funded_amnt              39717 non-null  int64  
 4   funded_amnt_inv          39717 non-null  float64
 5   term                     39717 non-null  object 
 6   int_rate                 39717 non-null  object 
 7   installment              39717 non-null  float64
 8   grade                    39717 non-null  object 
 9   sub_grade                39717 non-null  object 
 10  emp_title                37258 non-null  object 
 11  emp_length               38642 non-null  object 
 12  home_ownership           39717 non-null  object 
 13  annual_inc               39717 non-null  float64
 14  verification_status   

In [131]:
#Identify the columns whcih need to be in Date Fomrat
#issue_d
 #earliest_cr_line

#last_pymnt_d
#last_credit_pull_d

#Identify whihc clumsn need to be in int format

#term need to be converted to mumeric
#int_rate     remove %
#emp_length               need to be in numeric

#revol_util

In [132]:
#function to covert the object to date column

def toDate(cObj):
  x=cObj
  return x



In [133]:
# Identify columns where the percentage of null values is greater than 40%

null_percentage = df1.isnull().mean() * 100
cnd = null_percentage[(null_percentage < 40) & (null_percentage != 0)].index


df1[cnd].isnull().mean()*100

emp_title               6.191303
emp_length              2.706650
revol_util              0.125891
last_pymnt_d            0.178765
last_credit_pull_d      0.005036
pub_rec_bankruptcies    1.754916
dtype: float64

Review these null values at End #404

emp_title                6.191303
desc                    32.585543
title                    0.027696

last_pymnt_d             0.178765
last_credit_pull_d       0.005036
pub_rec_bankruptcies     1.754916


NOt able to decide is we can drop or not


In [134]:
#emp_lenght is  2.706650 null. Filling it with Median
#revol_util               0.125891
#examine the data in the column

df1.emp_length.unique()
df1.emp_length.nunique()
df1.emp_length.value_counts()


emp_length
10+ years    8879
< 1 year     4583
2 years      4388
3 years      4095
4 years      3436
5 years      3282
1 year       3240
6 years      2229
7 years      1773
8 years      1479
9 years      1258
Name: count, dtype: int64

In [135]:
# Convert Object to int

def removeLeftChar(input, numofchar):
  ans =input[numofchar:]
  return ans


def removeRighttChar(input, numofchar):
  input = str(input)
  ans =input[:len(input)-numofchar]

  return ans


def removeLastChar(input):
  ans = removeRighttChar(input,1)
  return ans

print(removeLeftChar(">1",1))
print(removeRighttChar("50%",1))



def getYears(x):
  if isinstance(x, str):
    xlist=x.split(" ")
  else:
    xlist = []


  if(len(xlist) == 3):
    ans=0
  elif (len(xlist) == 2):
    ans=int(x.split(" ")[0][:2])
  elif pd.isna(x):
    ans=x
  else:
      ans=int(x)
  return ans





1
50


In [136]:

# Convert object columns to datetime
date_columns = ['issue_d', 'earliest_cr_line', 'last_pymnt_d', 'last_credit_pull_d']
df1[date_columns] = df1[date_columns].apply(pd.to_datetime, format='%b-%y')


In [137]:
# Convertng Object to int
df1['emp_length'] =  df1['emp_length'].apply(getYears)

In [138]:
#int_rate     remove %
df1['int_rate'] =   df1['int_rate'].apply(removeLastChar).astype(int)

#int_rate     remove %
df1['revol_util'] =  df1['revol_util'].apply(removeLastChar).astype(int)

ValueError: invalid literal for int() with base 10: '10.65'

In [ ]:
df1['revol_util'].head()
df1.dtypes

In [ ]:
df1.emp_length.unique()
df1.emp_length.value_counts()

In [ ]:
#Find the median of the emp_length
import math

expmedian= math.trunc(int(df1.emp_length.median()))
print(math.trunc(expmedian))

df1.emp_length.isnull().sum()

df1.emp_length=df1.emp_length.fillna(math.trunc(expmedian))
df1.emp_length.value_counts()

In [ ]:
df1.isnull().mean()*100

In [ ]:
#removing columns as we might not be using it.
#Check why there is an error commenting for now
#df1=df1.drop(['mths_since_last_record','next_pymnt_d'],axis=1)

## **Univariate Analysis**

* Identify the columns to be Analyzed
* Generate plots ( Bar , Box Plot and Histogram )
* Review each column and *remove outliers


##Segmentation

In [ ]:
df1.head(3)

In [ ]:
df1.columns

In [ ]:
cat_cols=[ 'grade', 'sub_grade','home_ownership', 'verification_status',
       'issue_d', 'loan_status', 'pymnt_plan','zip_code', 'addr_state','open_acc', 'pub_rec',]
num_cols=['loan_amnt','funded_amnt','funded_amnt_inv', 'term', 'int_rate', 'installment',
          'emp_length',  'annual_inc','dti', 'delinq_2yrs','inq_last_6mths', 'mths_since_last_delinq',
          'revol_bal','revol_util', 'total_acc','out_prncp',
          'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv','total_rec_prncp',
          'total_rec_int', 'total_rec_late_fee', 'recoveries','collection_recovery_fee', ]
extra_cols=['id', 'member_id',  'emp_title','url', 'desc', 'purpose',
       'title','earliest_cr_line',   'initial_list_status', 'last_pymnt_d', 'last_pymnt_amnt',
       'next_pymnt_d', 'last_credit_pull_d', 'collections_12_mths_ex_med']

### Numerical Columns Analysis

In [ ]:
for i in num_cols:
    sns.histplot(x=df1[i])
    plt.show()

In [ ]:
#collection recovery fee analysis
df1.collection_recovery_fee.value_counts()  #collection_recovery_fee

In [ ]:
sns.histplot(x=df1['collection_recovery_fee'], bins = range(0,10,1) )
plt.show()

## **Bivariate Analysis**

* Identify the pairs of columns to be Analyzed
* Generate cattered plots plots and identify the dependency


## **Multivariate Analysis**

## For all Numeric columns plot a heat map and identify the corelation


* *Correlation represents strength of relationship between variables*
* *Correlation lies in the range of -1 to 1*
* *Negative correlation  (0 to -1) shows inverse relationship*
* *Positive correlation (0 to 1) shows direct relationship*
* *0 correlation represents no relationship. *italicized text*

## Univariate Analysis --> Analysis with Single variable
--> Mean, Median, Max, Min, Std, Variance, Count
--> Distribution (Histogram,Distplot,Countplot,boxplot)
## Bivaraite Analsyis --> Analysis with two variables
--> Relationship between two variables (Scatterplot, boxplot,barplot etc)
## Multivariate Analysis
--> Relationship between more than variables (Heatmap etc)

Coomit at 12.15

Commiting at 12.10